proposals/p2360.md

```diff
@@ -177,19 +221,161 @@ fn F(MyType:! MyInterface, x: MyType) -> auto {
-`MyType:! MyInterface` instead of as `MyType:! Type`, then of course `MyType`
-and `MyType as MyInterface` are equivalent.
+`MyType:! MyInterface` instead of as `MyType:! type`, then `MyType` and
+`MyType as MyInterface` are equivalent.
+
+#### Impl lookup
+
+Prior to this proposal, when a member access names a member of interface `I`,
+the rules for `impl` lookup also depend on whether the left hand operand is a
+type:
+
+-   If the left hand operand is a type-of-type (now, facet type), `impl` lookup
+    is not performed.
+-   If the left hand operand evaluates to a type `T`, then `impl` lookup is
+    performed for `T as I`.
+-   Otherwise, the first operand is an expression of some type `T`, and `impl`
+    lookup is performed for `T as I`.
+
+Due to the non-uniform treatment of types and non-types, this rule is also
+changed to the following:
+
+-   For a simple member access `a.b` where `b` names a member of an interface
+    `I`:
+    -   If the interface member was found by searching a non-facet-type scope
+        `T`, for example a class or an adapter, then `impl` lookup is performed
+        for `T as I`. For example:
+        -   `MyClass.AliasForInterfaceMember` finds the member of the
+            `impl MyClass as I`, not the interface member.
+        -   `my_value.AliasForInterfaceMember`, with `my_value: MyClass`, finds
+            the member of the `impl MyClass as I`, and performs instance binding
+            if the member is an instance member.
+        -   In the case where the member was found in a base class of the class
+            that was searched, `T` is the derived class that was searched, not
+            the base class in which the name was declared.
+    -   Otherwise, `impl` lookup is not performed:
+        -   `MyInterface.AliasForInterfaceMember` finds the member in the
+            interface `MyInterface` and does not perform `impl` lookup.
+-   For a compound member access `a.(b)` where `b` names a member of an
+    interface `I`, `impl` lookup is performed for `T as I`, where:
+    -   If `b` is an instance member, `T` is the type of `a`.
+        -   `my_value.(Interface.InstanceInterfaceMember)()`, where `my_value`
+            is of type `MyClass`, uses `MyClass as Interface`.
+        -   `MyClass.(Interface.InstanceInterfaceMember)()` uses
+            `type as Interface`.
+    -   Otherwise, `a` is implicitly converted to `I`, and `T` is the result of
+        symbolically evaluating the converted expression.
+        -   `MyClass.(Interface.NonInstanceInterfaceMember)()` uses
+            `MyClass as Interface`.
+        -   `my_value.(Interface.NonInstanceInterfaceMember)()` is an error
+            unless `my_value` implicitly converts to a type.
+
+This approach aims to change as little as possible of the existing `impl` lookup
+behavior while removing inconsistencies and behavioral differences based on
+whether a value is a type.
```

Comment:
I think this is all good, except it doesn't seem like all the alternatives we considered are well covered.

We talked about this 11-10: https://docs.google.com/document/d/1tEt4iM6vfcY0O0DG0uOEMIbaXcZXlNREc2ChNiEtn_w/edit#
and in #syntax starting 11-07: https://discord.com/channels/655572317891461132/709488742942900284/1039298322625744987